src/wasm/WasmModule.cpp

```diff
@@ -430,69 +426,77 @@ std::future<int32_t> WasmModule::executeOpenMPTask(threads::OpenMPTask t)
     // Enqueue the task
     std::promise<int32_t> p;
     std::future<int32_t> f = p.get_future();
-    openMPTaskQueue.enqueue(std::make_pair(std::move(p), std::move(t)));
 
-    // Check if we can add a thread if we're not shutting down
-    if (!t.isShutdown && (openMPThreads.size() < threadPoolSize)) {
+    // We have a fixed pool of threads, we have to share the work evenly over
+    // them, but we always want to send the same OMP thread number to the same
+    // thread. However, if we don't have enough threads in the pool, we want to
+    // cycle round again, and if we're at a nested level, we want to spread the
+    // work to other threads so we don't get nested ones blocking each other.
+    int threadPoolIdx =
+      (t.nextLevel->depth - 1 + t.msg->ompthreadnum()) % threadPoolSize;
```

Comment:
Just to double-check, can `t.nextLevel->depth` ever be 0? This is, can:
```cpp
t.nextLevel->depth - 1 + t.msg->ompthreadnum()
```
ever be -1? I guess this would yield undefined behaviour.

Reply:
Yes it's a good point. I'm 90% sure the depth of the lowest level is `1`, but I will definitely check, and perhaps throw an exception if the depth is less than 1.